cmd/finch-daemon/main.go

```diff
@@ -104,32 +104,41 @@ func getListener(options *DaemonOptions) (net.Listener, error) {
 	return listener, nil
 }
 
-func handlePidFileOptions(options *DaemonOptions) error {
-	if options.pidFile != "" {
-		if err := os.MkdirAll(filepath.Dir(options.pidFile), 0o640); err != nil {
-			return fmt.Errorf("failed to create pidfile directory %s", err)
-		}
-		if err := pidfile.Write(options.pidFile, os.Getpid()); err != nil {
-			return fmt.Errorf("failed to start daemon, ensure finch daemon is not running or delete %s %w", options.pidFile, err)
-		}
-	}
-	return nil
-}
-
 func run(options *DaemonOptions) error {
 	// This sets the log level of the dependencies that use logrus (e.g., containerd library).
 	if options.debug {
 		logrus.SetLevel(logrus.DebugLevel)
 	}
 
-	defer func() {
-		if err := os.Remove(options.pidFile); err != nil {
-			logrus.Errorf("failed to remove pidfile %s", options.pidFile)
+	if options.pidFile != "" {
+		if err := os.MkdirAll(filepath.Dir(options.pidFile), 0o600); err != nil {
+			return fmt.Errorf("failed to create pidfile directory %s", err)
 		}
-	}()
 
-	if err := handlePidFileOptions(options); err != nil {
-		return err
+		pidFile, err := os.OpenFile(options.pidFile, os.O_RDWR|os.O_CREATE, 0o600)
+		if err != nil {
+			return fmt.Errorf("failed to open or create PID file: %w", err)
+		}
+
+		// Lock the PID file to prevent concurrent access
+		if err := syscall.Flock(int(pidFile.Fd()), syscall.LOCK_EX|syscall.LOCK_NB); err != nil {
+			pidFile.Close()
+			return fmt.Errorf("failed to acquire lock on PID file; ensure only one instance is running")
```

Comment:
We can have multiple instances, just not with the same pidFile path. Suggest changing the error to
```suggestion
			return fmt.Errorf("failed to acquire lock on PID file (%s); ensure only one instance is using the PID file path", options.pidFile)
```